internal/hcsoci/create.go

```diff
@@ -229,3 +245,84 @@ func CreateContainer(ctx context.Context, createOptions *CreateOptions) (_ cow.C
 	}
 	return system, resources, nil
 }
+
+// CloneContainer is similar to CreateContainer but it does not add layers or namespace like
+// CreateContainer does. Also, instead of sending create container request it sends a modify
+// request to an existing container. CloneContainer only works for WCOW.
+func CloneContainer(ctx context.Context, createOptions *CreateOptions) (_ cow.Container, _ *Resources, err error) {
+	coi, err := initializeCreateOptions(ctx, createOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if coi.Spec.Windows == nil || coi.HostingSystem == nil {
+		return nil, nil, fmt.Errorf("CloneContainer is only supported for Hyper-v isolated WCOW ")
+	}
+
+	resources := &Resources{
+		id: createOptions.ID,
+	}
+	defer func() {
+		if err != nil {
+			if !coi.DoNotReleaseResourcesOnFailure {
+				ReleaseResources(ctx, resources, coi.HostingSystem, true)
+			}
+		}
+	}()
+
+	if coi.HostingSystem != nil {
+		n := coi.HostingSystem.ContainerCounter()
+		if coi.Spec.Linux != nil {
+			resources.containerRootInUVM = fmt.Sprintf(lcowRootInUVM, createOptions.ID)
+		} else {
+			resources.containerRootInUVM = fmt.Sprintf(wcowRootInUVM, strconv.FormatUint(n, 16))
+		}
+	}
+
+	if err = setupMounts(ctx, coi, resources); err != nil {
+		return nil, resources, err
+	}
+
+	// everything that is added to the container during the createContainer request
+	// (via the gcsDocument) must be hot added here.  Add the mounts as mapped
+	// directories or mapped pipes. In case of cloned container we must add them as a
+	// modify request.
+	_, _, mdsv2, mpsv2, err := createMountsConfig(ctx, coi)
+	if err != nil {
+		return nil, resources, err
+	}
+
+	c, err := coi.HostingSystem.CloneContainer(ctx, coi.actualID, nil)
+	if err != nil {
+		return nil, resources, err
+	}
+
+	// send modify requests for mounts one by one.
+	// TODO(ambarve) : Find out if there is a way to send request for all the mounts
+	// at the same time to save time
+	for _, md := range mdsv2 {
+		requestDocument := &hcsschema.ModifySettingRequest{
+			RequestType:  requesttype.Add,
+			ResourcePath: "Container/MappedDirectories",
+			Settings:     md,
+		}
+		err := c.Modify(ctx, requestDocument)
+		if err != nil {
+			return c, resources, fmt.Errorf("Error while adding mapped directory (%s) to the container: %s", md.HostPath, err)
+		}
+	}
+
+	for _, mp := range mpsv2 {
+		requestDocument := &hcsschema.ModifySettingRequest{
+			RequestType:  requesttype.Add,
+			ResourcePath: "Container/MappedPipes",
+			Settings:     mp,
+		}
+		err := c.Modify(ctx, requestDocument)
```

Comment:
Can we move these modify requests to their own files the way we do for uvm device modify requests? Then add the resource paths as globals?